File: Array/split_odd_even.py
from random import randint
from time import time

def split_odd_even(arr):
    """Split an array into two parts such that the first part (in front of the array)
        contain all odd nums, and the second part(in rear of the array) contains all
        even nums.
       arr: a list of integers to split, and it will be modified.
       return None.
       algorithm: use two indices, i and j, to traverse the array from its
         two ends towards its center, and swap arr[i], arr[j] when arr[i]
         is even and arr[j] is odd.
    """

    if None == arr or len(arr) < 2:
        return None
    assert isinstance(arr,list)
    i,j = 0,len(arr)-1
    while i < j:
        while i < j and 1 == arr[i] % 2:
            i += 1
        while i < j and 0 == arr[j] % 2:
            j -= 1
        arr[i],arr[j] = arr[j], arr[i]

def test():
    mn,mx = -13,13
    ln = 23
    arr = [randint(mn,mx) for _ in range(ln)]
    print(arr)
    split_odd_even(arr)
    print(arr)

if __name__ == '__main__':
    test()

File: Array/split_odd_even.cpp
#include <ctime>
#include <cstdlib>
#include <algorithm>
#include <iostream>
using namespace std;

/*
Split an array into two parts such that the first part (in front of the array)
contain all odd nums, and the second part(in rear of the array) contains all
even nums.
algorithm: use two indices, i and j, to traverse the array from its
         two ends towards its center, and swap arr[i], arr[j] when arr[i]
         is even and arr[j] is odd.

*/
template <typename T>
void split_odd_even(T arr[],int len)
{
    cout << endl << "len: " << len << endl;

    if (len < 2)
        return;
    int i, j;
    i = 0;
    j = len -1;
    while (i < j)
    {
        while(i < j && (1 == abs(arr[i] % 2)))
            ++ i;
        while (i < j && (0 == abs(arr[j] % 2)))
            -- j;
        swap(arr[i], arr[j]);
    }
}

void test()
{
    int *arr = NULL, len = 23;
    int min  = -13;
    int mx   = 33;
    arr      = new int[len];
    if (NULL == arr)
    {
        cout << "fail to new an array!" << endl;
        exit(-1);
    }
    srand(time(NULL));
    for (int i = 0; i < len; ++i)
        arr[i] = rand() % mx + min;
        // cout << rand() << endl;
    cout << "original arr: " << endl;
    for (int i = 0; i < len; ++i)
        cout << arr[i] << " ";
    split_odd_even(arr,len);
    cout << "splited: " << endl;
    for (int i = 0; i < len; ++i)
        cout << arr[i] << " ";
    cout << endl;
}

int main(int argc, char const *argv[])
{
    test();
    return 0;
}